tools/azure-devtools/src/azure_devtools/scenario_tests/preparers.py

```diff
@@ -5,17 +5,28 @@
 
 import contextlib
 import functools
+import logging
+import sys
+from threading import Lock
 
 from .base import ReplayableTest
 from .utilities import create_random_name, is_text_payload, trim_kwargs_from_test_function
 from .recording_processors import RecordingProcessor
 from .exceptions import AzureNameError
 
+_logger = logging.getLogger("preparer")
+_logger.setLevel(logging.DEBUG)
+handler = logging.StreamHandler(stream=sys.stdout)
+handler.setFormatter(logging.Formatter('%(asctime)s %(name)-12s %(levelname)-8s %(message)s'))
+_logger.addHandler(handler)
 
 # Core Utility
 
 
 class AbstractPreparer(object):
+    _cache_lock = Lock()
+    _resource_cache = {}
+    _cache_pending_deletes = {}
```

Comment:
Being you clean using "values()", order won't be respected. I would more imagine stack for LIFO behavior

Reply:
a queue complicates the process of deletion, although I could _probably_ be a bit more precise with only adding to the deletion cache on first creation which would maintain the "only once" nature.   Will try to clean this up++